python_modules/dagster/dagster/_utils/caching_instance_queryer.py

```diff
@@ -373,17 +378,21 @@ def next_version_record(
         after_cursor: Optional[int],
         data_version: Optional[DataVersion],
     ) -> Optional["EventLogRecord"]:
-        for record in self.instance.get_event_records(
-            EventRecordsFilter(
-                event_type=DagsterEventType.ASSET_OBSERVATION,
-                asset_key=asset_key,
-                after_cursor=after_cursor,
-            ),
-            ascending=True,
-        ):
-            record_version = extract_data_version_from_entry(record.event_log_entry)
-            if record_version is not None and record_version != data_version:
-                return record
+        has_more = True
+        cursor = None
+        while has_more:
+            result = self.instance.fetch_observations(
+                AssetRecordsFilter(asset_key=asset_key, after_storage_id=after_cursor),
+                limit=RECORD_BATCH_SIZE,
```

Comment:
Are we "double batching" here? I.e. should we not trust the `fetch_observations` implementation to determine the right batch size?

Reply:
I'm not sure what you mean by this.

Reply:
The implementation of `fetch_observations` internally sets a limit on how many records that it fetches from the DB at once, right? I.e. this is what `has_more` is for, so that the implementation doesn't need to return all the records at once?

And then `RECORD_BATCH_SIZE` is presumably also setting a limit on how many records to fetch at once.